File: plugin/octo_beat_plugin/android/src/main/kotlin/com/octo/octo_beat_plugin/core/utils/PacketLengthHelper.java
package com.octo.octo_beat_plugin.core.utils;

/**
 * Created by caoxuanphong on 12/18/17.
 */

public class PacketLengthHelper {

    public static int encode_length(int len, byte[] buf) {
        int encodedByte, cnt;

        assert (len <= 0xFFFFFFF);
        cnt = 0;

        do {
            encodedByte = len & 0x7F;
            len >>= 7;
            encodedByte |= (len != 0) ? 0x80 : 0x00;

            buf[cnt++] = (byte) encodedByte;
        } while (len != 0);

        return cnt;
    }

    public static int decodePacketLength(byte[] buf, int size) {
        int multiplier, value;
        int decodedByte;
        int cnt = 0;

        assert (size <= 4);
        assert ((buf[size - 1] & 0x80) == 0); // The MSb of the last byte must be '0'

        value = 0;
        multiplier = 1;

        do {
            decodedByte = buf[cnt++];
            value += (decodedByte & 0x7F) * multiplier;
            multiplier <<= 7;
        } while (--size > 0);

        return value;
    }
}

File: plugin/octo_beat_plugin/ios/Classes/core/utils/PacketLengthHelper.swift

//
//  PacketLengthHelper.swift
//  DXH-ios
//
//  Created by Manh Tran on 11/22/19.
//  Copyright © 2019 software-2. All rights reserved.
//

import Foundation
public class PacketLengthHelper {
    public static func encodeLength( len: inout UInt32, buf: inout [UInt8]) -> UInt8{
        var encodeByte, cnt : UInt8
        assert(len <= 0xFFFFFFF)
        cnt = 0
        
        repeat {
            encodeByte = UInt8(len & 0x7F)
            len >>= 7
            encodeByte |= (len != 0) ? 0x80 : 0x00
            cnt += 1
            buf.append(UInt8(encodeByte))
            
        } while(len != 0)
        
        return cnt
    }
    
    public static func decodePakcetLength(buf: [UInt8], size: inout Int) -> UInt32{
        var mutiplier, value: UInt32
        var decodeByte: UInt8
        var cnt = 0
        
        assert(size <= 4)
        assert((buf[size - 1] & 0x80) == 0)
        
        value = 0
        mutiplier = 1
        repeat {
            decodeByte = buf[cnt]
            value += UInt32((decodeByte & 0x7F)) * mutiplier
            mutiplier <<= 7
            cnt += 1
            size -= 1
        } while(size > 0)
        
        return value
    }
}
